src/api/controllers/AuthController.ts

```diff
@@ -43,22 +56,56 @@ export class AuthController {
   async login(
     @Req() req: Request,
     @Res() res: Response,
-    // Just used for validation (email and password are in passport strategy)
+    /**
+     * `email` and `password` just used for validation (email and password are in passport strategy)
+     * `token` is used for multi factor authentication
+     */
     @Body() data: LoginData
   ): Promise<void> {
-    await new Promise<Contact>((resolve, reject) => {
-      passport.authenticate("local", (err, user, info) => {
-        if (err || !user) {
-          const error = new UnauthorizedError() as any;
-          if (info?.message) {
-            error.code = info.message;
+    const user = await new Promise<Contact>((resolve, reject) => {
+      passport.authenticate(
+        "local",
+        async (err: null, user: Contact | false, info?: PassportLoginInfo) => {
+          if (err || !user) {
+            // Unknown error
+            return reject(
+              new UnauthorizedError({
+                code: info?.message || LOGIN_CODES.LOGIN_FAILED
+              })
+            );
           }
-          reject(error);
-        } else {
+
+          // If user has 2FA enabled, check token
+          if (info?.message === LOGIN_CODES.REQUIRES_2FA) {
+            // If user has no token, notify client that 2FA is required
+            if (!data.token) {
```

Comment:
I think all this code should be within the local passport strategy?

Reply:
Yes I thought so too, problem is that the `passport-local` module only has support for email and password and so I can't get the token, at this point I got the token from the controller outside the callback.

One solution would be to implement a new passport strategy (as suggested at the beginning) or to fork the passport-local strategy.

Reply:
Resolving this thread to keep our discussion in one place